src/vm.c

```diff
@@ -72,6 +76,87 @@ static void vm_init_cpu_id(vm_t *v)
     ioctl(v->vcpu_fd, KVM_SET_CPUID2, &kvm_cpuid);
 }
 
+#define FREQ_NS 1000000
+
+/* global state to stop the loop of thread */
+static volatile bool thread_stop = false;
+static pthread_t main_thread;
+
+static void *timer_thread(void *args)
+{
+    struct epoll_event events[1];
+    int epoll_fd = *(int *) args;
+
+    while (!__atomic_load_n(&thread_stop, __ATOMIC_RELAXED)) {
+        int nfds = epoll_wait(epoll_fd, events, 1, -1);
+        if (nfds < 0)
+            continue;
+
+        assert(nfds == 1);
+
+        uint64_t exp;
+        ssize_t size = read(events[0].data.fd, &exp, sizeof(uint64_t));
+        if (size != sizeof(uint64_t))
+            return NULL;
+
+        pthread_kill((pthread_t) main_thread, SIGUSR1);
+    }
+
+    return NULL;
+}
+
+static void handler(int sig, siginfo_t *si, void *uc) {}
+
+static int vm_init_timer(vm_t *v)
+{
+    int epoll_fd, timer_fd;
+    struct epoll_event ev;
+    struct itimerspec its;
+    struct sigaction sa;
+    sigset_t mask;
+
+    sa.sa_flags = SA_SIGINFO;
+    sa.sa_sigaction = handler;
+    sigemptyset(&sa.sa_mask);
+    if (sigaction(SIGUSR1, &sa, NULL) == -1)
+        return throw_err("Failed to create signal handler");
+
+    /* Block timer signal temporarily. */
+    sigemptyset(&mask);
+    sigaddset(&mask, SIGUSR1);
```

Comment:
Since `SIGUSR1` is introduced for flexible flow control, I just wonder if we can remove the use of POSIX Threads completely.

Reply:
The key point for the serial input feature: interrupt the main thread. This is not simple

Taking [kvmtool](https://github.com/kvmtool/kvmtool/blob/master/term.c) for example, it also creates a POSIX thread. But instead of looking at the timerfd, it looks at the STDIN or something similarly directly (I think this sounds like a better idea! How about changing to this design?)

Reply:
Did you mean the way to use `epoll` to listen for STDIN events? It sounds straightforward and efficient.

Reply:
Yes, instead of listening to the timerfd to interrupt the main thread periodically, `kvmtool` listens to STDIN events and interrupts the main thread only when there do really have keyboard inputs. I think the latter implementation is better and asking if we want to change for that.

Reply:
Although we still need a thread for this implementation, now the thread doesn't work on `vm_t`. It just tells the main thread that it needs to spend some time for the serial inputs first. So we don't need to worry about data race on this design.

Reply:
The advantage of using `epoll` to listen for STDIN events:
* Compare to the design of kvm-host on master branch, we can remove the mutex
* Compare to the design of the original pull request, we can reduce the useless interrupt on the main thread

Reply:
Affirmative! Please work on listening for STDIN events.